File: Habitica/src/main/java/com/habitrpg/android/habitica/receivers/NotificationPublisher.java
package com.habitrpg.android.habitica.receivers;

import android.annotation.TargetApi;
import android.app.Notification;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.support.v4.content.ContextCompat;
import android.support.v4.content.WakefulBroadcastReceiver;

import com.habitrpg.android.habitica.HabiticaApplication;
import com.habitrpg.android.habitica.R;
import com.habitrpg.android.habitica.data.TaskRepository;
import com.habitrpg.android.habitica.helpers.RxErrorHandler;
import com.habitrpg.android.habitica.helpers.TaskAlarmManager;
import com.habitrpg.android.habitica.models.tasks.Task;
import com.habitrpg.android.habitica.modules.AppModule;
import com.habitrpg.android.habitica.ui.activities.MainActivity;

import javax.inject.Inject;
import javax.inject.Named;


//https://gist.github.com/BrandonSmith/6679223
public class NotificationPublisher extends WakefulBroadcastReceiver {

    @Inject
    TaskRepository taskRepository;
    @Inject
    @Named(AppModule.NAMED_USER_ID)
    String userId;

    public static String NOTIFICATION_ID = "notification-id";
    public static String CHECK_DAILIES = "check-dailies";
    private Context context;

    public void onReceive(Context context, Intent intent) {
        this.context = context;
        if (taskRepository == null) {
            HabiticaApplication.getComponent().inject(this);
        }

        boolean check_dailies = intent.getBooleanExtra(CHECK_DAILIES, false);
        Notification notification = getNotification();
        if (check_dailies) {
            taskRepository.getTasks(Task.TYPE_DAILY, userId).subscribe(dailies -> {
                boolean showNotifications = false;
                for (Task task : dailies) {
                    if (task.checkIfDue(0)) {
                        showNotifications = true;
                        break;
                    }
                }
                TaskAlarmManager.scheduleDailyReminder(context);
                if (showNotifications) {
                    NotificationManager notificationManager = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);

                    int id = intent.getIntExtra(NOTIFICATION_ID, 0);
                    notificationManager.notify(id, notification);
                }
            }, RxErrorHandler.handleEmptyError());

        } else {
            NotificationManager notificationManager = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);

            int id = intent.getIntExtra(NOTIFICATION_ID, 0);
            notificationManager.notify(id, notification);
        }
    }

    @TargetApi(21)
    private Notification getNotification() {
        int currentapiVersion = android.os.Build.VERSION.SDK_INT;

        Notification notification;
        Notification.Builder builder = new Notification.Builder(context);
        builder.setContentTitle(context.getString(R.string.app_name));
        builder.setContentText(context.getString(R.string.reminder_title));
        builder.setSmallIcon(R.drawable.ic_gryphon_white);
        Intent notificationIntent = new Intent(context, MainActivity.class);

        notificationIntent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP
                | Intent.FLAG_ACTIVITY_SINGLE_TOP);

        PendingIntent intent = PendingIntent.getActivity(context, 0,
                notificationIntent, 0);
        builder.setContentIntent(intent);

        if (currentapiVersion >= Build.VERSION_CODES.LOLLIPOP) {
            builder.setColor(ContextCompat.getColor(context, R.color.brand_300));
        }

        if (currentapiVersion >= Build.VERSION_CODES.JELLY_BEAN) {
            notification = builder.build();
        } else {
            notification = builder.getNotification();
        }
        notification.defaults |= Notification.DEFAULT_LIGHTS;

        notification.flags |= Notification.FLAG_AUTO_CANCEL | Notification.FLAG_SHOW_LIGHTS;
        return notification;
    }
}

File: Habitica/src/main/java/com/habitrpg/android/habitica/receivers/NotificationPublisher.kt
package com.habitrpg.android.habitica.receivers

import android.app.Notification
import android.app.PendingIntent
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import androidx.core.content.ContextCompat
import androidx.core.content.edit
import com.habitrpg.android.habitica.HabiticaBaseApplication
import com.habitrpg.android.habitica.R
import com.habitrpg.android.habitica.data.TaskRepository
import com.habitrpg.android.habitica.data.UserRepository
import com.habitrpg.android.habitica.extensions.withImmutableFlag
import com.habitrpg.android.habitica.helpers.TaskAlarmManager
import com.habitrpg.common.habitica.helpers.launchCatching
import com.habitrpg.android.habitica.ui.activities.MainActivity
import com.habitrpg.shared.habitica.models.tasks.TaskType
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.flow.firstOrNull
import java.util.Calendar
import java.util.Date
import java.util.Random
import javax.inject.Inject

// https://gist.github.com/BrandonSmith/6679223
class NotificationPublisher : BroadcastReceiver() {

    @Inject
    lateinit var taskRepository: TaskRepository
    @Inject
    lateinit var userRepository: UserRepository
    @Inject
    lateinit var sharedPreferences: SharedPreferences

    private var wasInjected = false
    private var context: Context? = null

    override fun onReceive(context: Context, intent: Intent) {
        this.context = context
        if (!wasInjected) {
            wasInjected = true
            HabiticaBaseApplication.userComponent?.inject(this)
        }

        var wasInactive = false
        // Show special notification if user hasn't logged in for a week
        if (sharedPreferences.getLong("lastAppLaunch", Date().time) < (Date().time - 604800000L)) {
            wasInactive = true
            sharedPreferences.edit { putBoolean("preventDailyReminder", true) }
        } else {
            TaskAlarmManager.scheduleDailyReminder(context)
        }
        val checkDailies = intent.getBooleanExtra(CHECK_DAILIES, false)
        if (checkDailies) {
            MainScope().launchCatching {
                val tasks = taskRepository.getTasks(TaskType.DAILY, null, emptyArray()).firstOrNull()
                val user = userRepository.getUser().firstOrNull()
                var showNotifications = false
                for (task in tasks ?: emptyList()) {
                    if (task.checkIfDue()) {
                        showNotifications = true
                        break
                    }
                }
                if (showNotifications) {
                    notify(intent, buildNotification(wasInactive, user?.authentication?.timestamps?.createdAt))
                }
            }
        } else {
            notify(intent, buildNotification(wasInactive))
        }
    }

    private fun notify(intent: Intent, notification: Notification?) {
        val notificationManager = context?.let { NotificationManagerCompat.from(it) }
        val id = intent.getIntExtra(NOTIFICATION_ID, 0)
        notification?.let { notificationManager?.notify(id, it) }
    }

    private fun buildNotification(wasInactive: Boolean, registrationDate: Date? = null): Notification? {
        val thisContext = context ?: return null
        val notification: Notification
        val builder = NotificationCompat.Builder(thisContext, "default")
        builder.setContentTitle(thisContext.getString(R.string.reminder_title))
        var notificationText = getRandomDailyTip()
        if (registrationDate != null) {
            val registrationCal = Calendar.getInstance()
            registrationCal.time = registrationDate
            val todayCal = Calendar.getInstance()
            val isSameDay = (
                registrationCal.get(Calendar.YEAR) == todayCal.get(Calendar.YEAR) &&
                    registrationCal.get(Calendar.DAY_OF_YEAR) == todayCal.get(Calendar.DAY_OF_YEAR)
                )
            val isPreviousDay = (
                registrationCal.get(Calendar.YEAR) == todayCal.get(Calendar.YEAR) &&
                    registrationCal.get(Calendar.DAY_OF_YEAR) == (todayCal.get(Calendar.DAY_OF_YEAR) - 1)
                )
            if (isSameDay) {
                builder.setContentTitle(thisContext.getString(R.string.same_day_reminder_title))
                notificationText = thisContext.getString(R.string.same_day_reminder_text)
            } else if (isPreviousDay) {
                builder.setContentTitle(thisContext.getString(R.string.next_day_reminder_title))
                notificationText = thisContext.getString(R.string.next_day_reminder_text)
            }
        }
        if (wasInactive) {
            builder.setContentText(thisContext.getString(R.string.week_reminder_title))
            notificationText = thisContext.getString(R.string.week_reminder_text)
        }

        builder.setStyle(NotificationCompat.BigTextStyle().bigText(notificationText))

        builder.setSmallIcon(R.drawable.ic_gryphon_white)
        val notificationIntent = Intent(thisContext, MainActivity::class.java)
        notificationIntent.putExtra("notificationIdentifier", "daily_reminder")

        notificationIntent.flags = Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_SINGLE_TOP

        val intent = PendingIntent.getActivity(
            thisContext, 0,
            notificationIntent, withImmutableFlag(0)
        )
        builder.setContentIntent(intent)

        builder.color = ContextCompat.getColor(thisContext, R.color.brand_300)

        notification = builder.build()
        notification.defaults = notification.defaults or Notification.DEFAULT_LIGHTS

        notification.flags = notification.flags or (Notification.FLAG_AUTO_CANCEL or Notification.FLAG_SHOW_LIGHTS)
        return notification
    }

    private fun getRandomDailyTip(): String {
        val thisContext = context ?: return ""
        return when (Random().nextInt(10)) {
            0 -> thisContext.getString(R.string.daily_tip_0)
            1 -> thisContext.getString(R.string.daily_tip_1)
            2 -> thisContext.getString(R.string.daily_tip_2)
            3 -> thisContext.getString(R.string.daily_tip_3)
            4 -> thisContext.getString(R.string.daily_tip_4)
            5 -> thisContext.getString(R.string.daily_tip_5)
            6 -> thisContext.getString(R.string.daily_tip_6)
            7 -> thisContext.getString(R.string.daily_tip_7)
            8 -> thisContext.getString(R.string.daily_tip_8)
            9 -> thisContext.getString(R.string.daily_tip_9)
            else -> ""
        }
    }

    companion object {
        var NOTIFICATION_ID = "notification-id"
        var CHECK_DAILIES = "check-dailies"
    }
}
